cores/esp8266/heap.cpp

```diff
@@ -90,96 +126,115 @@ static const char oom_fmt[]   PROGMEM STORE_ATTR = ":oom(%d)@?\n";
 static const char oom_fmt_1[] PROGMEM STORE_ATTR = ":oom(%d)@";
 static const char oom_fmt_2[] PROGMEM STORE_ATTR = ":%d\n";
 
-void* malloc (size_t s)
+void ICACHE_RAM_ATTR print_loc(size_t size, const char* file, int line)
 {
-    void* ret = umm_malloc(s);
-    if (!ret)
-        os_printf(oom_fmt, (int)s);
-    return ret;
+        DBGLOG_FUNCTION_P(oom_fmt_1, (int)size);
```

Comment:
Is this a question?

Reply:
Without looking at other code, I think that when debug port is disabled, dbglog_function should resolve to nothing, so this print_loc should get optimized away.

Reply:
Okay, I didn't go deep enough into this description. And sorry, I sometimes forget my question marks.

I wanted to trying and replicate the general debug logic flow of the original `heap.cpp`; however, I wasn't sure of the intent. Things get a little strange when Debug port: "disabled" is selected and Debug Level: "OOM" is selected. 

`os_printf` aka `os_printf_plus` was used to do the printing for OOM; however, because we have Debug port: "disabled" `system_set_os_print(1);` is not automatically called. Without this call `os_printf_plus` just exits, never printing. On the other hand it someone added a call to 
 `system_set_os_print(1);` or `Serial.setDebugOutput(true)` to their sketch, that would also have enabled printing. I am not sure anyone was aware of all this?

If this is the desired behavior I'll put `os_printf` back?
>A side note on os_printf (os_printf_plus). To handle printing PROGMEM, it will do a malloc for a buffer. If that fails it has a 64 character buffer in the stack it will use to copy the format string to. All of our current OOM strings will fit in the later.

If we go back to using `os_printf`, the ISR call path issue will need to be addressed. Probably needs to be anyway. Especially if the SDK blob is calling to print debug stuff from an ISR. (unconfirmed)

Reply:
@d-a-v were you aware of this? What do you think?

Reply:
I can't seem to find `DBGLOG_FUNCTION_P` definition. 
What I can say is since this code was written, `::printf()` from @earlephilhower 's libc is able to print PROGMEM strings without further allocation and I think is safe in ISRs. ref: #5728 
Need @earlephilhower 's confirmation though

Reply:
`::printf()` comes mostly unadulterated from newlib's nano-vprintf, with additions to support PROGMEM format and parameter strings (and a couple minor format additions from the main one to support additional strings).

It, unfortunately, can call an allocator (malloc, realloc) in the general case.  You should be able to check the specific format string used to determine if it will need one for any output you're generating, if you want to go that way.

Reply:
I remember only `"%s"` will issue a `malloc` (to be verified)

Reply:
@earlephilhower
To clarify:
* If just the format is in PROGMEM we are okay, no allocator?
* Is it safe to call from an ISR?

Reply:
@d-a-v here is the define for `DBGLOG_FUNCTION_P`
https://github.com/esp8266/Arduino/blob/57b65abfedf6559868ee7fa9c8533a32e7ca2a44/cores/esp8266/umm_malloc/umm_malloc_cfg.h#L59-L70
https://github.com/esp8266/Arduino/blob/71b10fdbf33982b55d9bec0eb21eea4397d9b3c9/cores/esp8266/umm_malloc/isr_safe_printf.cpp#L86-L96

Reply:
As to the memory allocation, that's a little complicated to figure out from manually call tracing.  In general it seems that only *asprintf* (allocating) printfs actually call the malloc/realloc to size a destination buffer.  Sources and format strings are built byte-by-byte and I do not believe they ever need allocation (at least not for non-FP types).

From ISR, it *should* be safe as behind the scenes it is the _r (reent) version.

Reply:
To be counter-verified:
I added a flag in all malloc functions, and this flag seems to be always toggled with `"%d"`, never with `"%s"`... (using `flag=0`, `::printf` with `PSTR("for fmts and args")`, `if (flag) print("(called)");`)

Reply:
@d-a-v  It seems to be a onetime thing, the 1st printf appears to do two mallocs. The last one is 140 bytes. After that, I don't see the malloc or realloc count more. I was thinking some kind of class instantiation; however, placing curly braces around the `::printf` does effect malloc stats.

@earlephilhower  It is good that it is entrant; however, when I look at the .map file for a build, I cannot any functions called printf that are in iram. What should I see?

Reply:
Ah, good check!  I normally only care about the postmortem, which has the ROM set flash enabled before being called so it's not a problem there, but I can see your point in an ISR.  I suggest it's best to stick with the limited version `ets_printf` in ROM for now.

Reply:
With `Debug Level: OOM` selected w/ `Debug port: "Disabled"` you can now use `Serial.setDebugOutput(true);` to enable `OOM` message displaying. This is the behavior previously available with `os_printf`.